vignettes/potential-optimizers.Rmd

```diff
@@ -0,0 +1,670 @@
+library("ggplot2")
+library("microbenchmark")
+n <- 10000
+autoplot(microbenchmark(non_vectorized(n), vectorized(n)))
+```
+
+### Drawbacks
+
+  **To-Do**
+
+## 5. Caching
+
+### Idea
+The idea of caching is similar to the concept of dynamic programming where the results of repeating sub-problems of a recursion are stored/cached and simply called every time instead of recomputating.
+
+### Code Examples
+
+#### Unoptimized Code
+
+```{r cache_code_og, echo=TRUE, warning=FALSE}
+
+library("memoise")
+
+fibR <- function(x) {
+  if (x == 0) return(0)
+  if (x == 1) return(1)
+  Recall(x - 1) + Recall(x - 2)
+}
+```
+
+#### Proposed Optimized Code
+
+```{r cache_code_op, echo=TRUE, warning=FALSE}
+mem_fibR <- memoise::memoise(fibR)
+```
+
+### Benchmark
+
+```{r cache_benchmark, warning=FALSE, message=FALSE}
+library("ggplot2")
+library("microbenchmark")
+n <- 20
+autoplot(microbenchmark(fibR(n), mem_fibR(n)))
+```
+
+### Drawbacks
+
+  **To-Do**
+
+## 6. Reversing elements
+
+### Idea
+Instead of nesting a function inside another function, it is generally faster to use a single more specific function. The `rev()` function provides a reversed version of its argument. If you wish to sort in decreasing order, `sort(x, decreasing = TRUE)` is around 10% faster than `rev(sort(x))`.
+
+### Code Examples
+
+#### Unoptimized Code
+
+```{r rev_code_og, echo = TRUE, warning = FALSE}
+reverse_fun <- function(n) {
+  vec <- sample(seq_len(100000), n)
+  rev(sort(vec))
+}
+```
+
+#### Proposed Optimized Code
+
+```{r rev_code_op, echo = TRUE, warning = FALSE}
+reverse_fun_opt <- function(n) {
+  vec <- sample(seq_len(100000), n)
+  sort(vec, decreasing = TRUE)
+}
+```
+
+### Benchmark
+
+```{r rev_benchmark, warning = FALSE, echo = TRUE, message = FALSE}
+library("ggplot2")
+library("microbenchmark")
+n <- 1000
+autoplot(microbenchmark(reverse_fun(n), reverse_fun_opt(n), times = 10))
+```
+
+### Drawbacks
+
+  **To-Do**
+  
+## 7. Efficient Sorting
+
+### Idea
+There are currently three sorting algorithms, `c("shell", "quick", "radix")` that can be specified in the `sort()` function; with `radix` being a new addition to `R 3.3`. Typically the `radix` **(the non-default option)** is the most computationally efficient option for most situations (it is around 20% faster when sorting a large vector of doubles)
+
+### Code Examples
+
+#### Unoptimized Code
+
+```{r sort_code_og, echo = TRUE, warning = FALSE}
+sort_fun <- function(n) {
+  vec <- sample(seq_len(1000000000), n)
+  sort(vec)
+}
+```
+
+#### Proposed Optimized Code
+
+```{r sort_code_op, echo = TRUE, warning = FALSE}
+
+sort_fun_opt <- function(n) {
+  vec <- sample(seq_len(100000000), n)
+  sort(vec, method = "radix")
+}
+```
+
+### Benchmark
+
+```{r sort_benchmark, warning = FALSE, echo = TRUE, message = FALSE}
+library("ggplot2")
+library("microbenchmark")
+n <- 100000
+autoplot(microbenchmark(sort_fun(n), sort_fun_opt(n)))
+```
+
+### Drawbacks
+
+  **To-Do**
+  
+## 8. min/max using which function
```

Comment:
Same as before, `rco` should focus on optimizing code chunks, not algorithms.
I mean, `rco` should reorganize code, resulting in more efficient code.
We should remove this idea for now.